environs/context/callcontext.go

```diff
@@ -60,11 +76,37 @@ func (a *legacyCredentialAdaptor) InvalidateModelCredential(reason string) error
 	return nil
 }
 
-// CallContext creates a CloudCallContext for use when calling environ methods
-// that may require invalidate a cloud credential.
-func CallContext(credential ModelCredentialInvalidator) *CloudCallContext {
-	// TODO(wallyworld) - pass in the stdcontext
-	callCtx := NewCloudCallContext(stdcontext.Background())
-	callCtx.InvalidateCredentialFunc = credential.InvalidateModelCredential
-	return callCtx
+// ProviderCallContext wraps a standard context
+// and is used in provider api calls.
+type ProviderCallContext struct {
+	stdcontext.Context
+}
+
+// WithoutCredentialInvalidator returns a ProviderCallContext
+// without any credential invalidation callback.
+func WithoutCredentialInvalidator(ctx stdcontext.Context) ProviderCallContext {
+	return ProviderCallContext{ctx}
+}
+
+const (
+	credentialInvalidatorKey = "credential-invalidator"
+)
+
+// WithCredentialInvalidator returns a ProviderCallContext
+// with the specified credential invalidation callback.
+func WithCredentialInvalidator(ctx stdcontext.Context, invalidationFunc InvalidateModelCredentialFunc) ProviderCallContext {
+	return ProviderCallContext{
+		stdcontext.WithValue(ctx, credentialInvalidatorKey, invalidationFunc),
```

Comment:
This is kind of an anti-pattern. Since it can affect behaviour (a part from cancellation), it is really an argument to the  functions being called. I think putting it back into the ProviderCallContext similar to it was before will correct this concern for me.

Reply:
I've added a struct attribute.